src/m3/constraints.ts

```diff
@@ -12,19 +19,37 @@ export const issuesMetamodel = (metamodel: Metamodel): Issue[] =>
     flatMap(
         metamodel,
         (t) => {
+
+            const issues: Issue[] = []
+            const issue = (message: string): void => {
+                issues.push({
+                    location: t,
+                    message
+                })
+            }
+
+            if (t instanceof Concept || t instanceof ConceptInterface) {
+                const cycle = inheritedCycleWith(t)
+                if (cycle.length > 0) {
+                    issue(`A ${t.constructor.name} can't inherit (directly or indirectly) from itself, but ${t.qualifiedName()} does so through the following cycle: ${cycle.map((t) => t.qualifiedName()).join(" -> ")}`)
+                }
+            }
+
             if (t instanceof ConceptInterface) {
                 const nonDerivedFeatures = t.allFeatures().filter(({derived}) => !derived)
                 if (nonDerivedFeatures.length > 0) {
                     const isPlural = nonDerivedFeatures.length > 1
-                    return [
-                        {
-                            location: t,
-                            message: `The features of a ConceptInterface must all be derived, but the following feature${isPlural ? `s` : ``} of ${t.simpleName} ${isPlural ? `are` : `is`} not: ${nonDerivedFeatures.map(({simpleName}) => simpleName).join(", ")}.`
-                        }
-                    ]
+                    issue(`The features of a ConceptInterface must all be derived, but the following feature${isPlural ? `s` : ``} of ${t.qualifiedName()} ${isPlural ? `are` : `is`} not: ${nonDerivedFeatures.map(({simpleName}) => simpleName).join(", ")}.`)
```

Comment:
Note: this part of having derived features only in ConceptInterface is a bit strange to me, but this is a problem for that decision, not for this implementation which is aligned with the decision we made during the meetings

Reply:
Indeed: I'm pretty sure it's something that we explicitly decided for.